ISPN-7127 Broken InboundTransferTask may lose segments on restart

danberindei opened the pull request:
https://issues.jboss.org/browse/ISPN-7127
- Only retry an inbound transfer when we receive a new cache topology.
- Move the segment list to the end of log messages

rvansa commented:
No test case? :-/

danberindei commented:
@rvansa Unfortunately I haven't been able to figure out how the random failures in `StateTransferRestartTest` actually happened, and my attempts to add strategic delays in `StateConsumerImpl` didn't make the failures easier to reproduce. I also looked at your `ConcurrentOverlappingLeaveTest` log, but I didn't see any failure there...

rvansa commented:
@danberindei IIRC in the ConcurrentOverlappingLeaveTest a segment was not transfered to a node since the transfer has been canceled (and new one did not start). But fair enough, let's hope that this PR actually fixes something... I hope I'll find some time to check the test log again, add more logging and catch it in pyjamas...

danberindei commented:
@rvansa I had added a new test while working on a more involved solution, but I decided not to add it to the PR because it didn't fail often enough. Weird enough, I tried it a few times on master today and it failed every time, so I've added it.

slaskawi commented on core/src/test/java/org/infinispan/statetransfer/StateTransferRestart2Test.java:
```diff
@@ -0,0 +1,154 @@
+
+   private ConfigurationBuilder cfgBuilder;
+
+   private void waitForStateTransfer(Cache... caches) throws InterruptedException {
```
Would it be possible to use `org.infinispan.test.TestingUtil#waitForRehashToComplete(org.infinispan.Cache...)` here?

danberindei replied:
I think so, I copied it from `StateTrasnsferRestartTest` without paying too much attention.

slaskawi commented on core/src/test/java/org/infinispan/statetransfer/StateTransferRestart2Test.java:
```diff
@@ -0,0 +1,154 @@
+      log.infof("c0 entrySet size before : %d", c0.entrySet().size());
+      log.infof("c2 entrySet size before : %d", c2.entrySet().size());
+
+      eventually(new Condition() {
```
This could be a lambda. It would be shorter.

wburns commented on core/src/main/java/org/infinispan/statetransfer/InboundTransferTask.java:
```diff
@@ -118,7 +118,7 @@ public Address getSource() {
          }
          if (trace) {
-            log.tracef("Requesting segments %s of cache %s from node %s", segmentsCopy, cacheName, source);
+            log.tracef("Requesting state from node %s for segments %s", source, segmentsCopy);
```
Sorry I might be missing something, but why take out the cacheName?  Seems we would lose some information in the log messages.  Or is it just information overload? :)

danberindei replied:
I think the cache name was already in the NDC. Also, none of the other messages contain the cache name.

wburns commented on core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java:
```diff
@@ -182,13 +182,13 @@ public void processEntry(MarshalledEntry me, AdvancedCacheLoader.TaskContext tas
          // ignore eventual exceptions caused by cancellation (have InterruptedException as the root cause)
          if (isCancelled()) {
-            log.debugf("Transfer of segments %s of cache %s to node %s cancelled", segments, cacheName, destination);
+            log.tracef("Ignoring error in already cancelled transfer to node %s, segments %s", destination, segments);
```
Looks like we want a `if (trace)` around this?

wburns commented on core/src/test/java/org/infinispan/statetransfer/StateTransferRestart2Test.java:
```diff
@@ -0,0 +1,154 @@
+         if (!inProgress) {
+            break;
+         }
+         wait(100);
```
It doesn't seem like we are holding the lock on `this` when calling wait, which is required as part of the spec.

danberindei replied:
I guess it never actually got to this point :) I'm removing this method and using `TestingUtil.waitForRehashToComplete()` instead.

galderz commented:
No @danberindei replies since October to @wburns' and @slaskawi's comments. Closing for now. @danberindei reopen when you have time to reply to them.

danberindei commented:
Oops, can't reopen after you force-push...
